apps/utils/lib/ledgers/wallet_schema.ex

```diff
@@ -0,0 +1,27 @@
+# Copyright 2017-2019 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule Utils.Ledgers.WalletSchema do
+  @moduledoc """
+  Defines common behaviours for ledger wallet schemas.
+  """
+
+  @callback all([String.t()]) :: [Ecto.Schema.t()]
+  @callback get(String.t()) :: Ecto.Schema.t() | nil
+  @callback get_by(Keyword.t() | map()) :: Ecto.Schema.t() | nil
+  @callback get_or_insert(map()) :: {:ok, Ecto.Schema.t()} | {:error, Ecto.Changeset.t()}
+  @callback insert(map()) :: {:ok, Ecto.Schema.t()} | {:error, Ecto.Changeset.t()}
+  @callback touch([String.t()]) :: {integer(), nil | [term()]}
+  @callback lock([String.t()]) :: [Ecto.Schema.t()]
```

Comment:
Not sure we need this, local wallets and ledger wallets are not supposed to work in the same way. You won't be able to lock a blockchain wallet (the balances are kept on the blockchain anyway, so the blockchain is responsible for preventing double spends). 

When making a transaction, a few scenarios can occur: 

- local to local (like now)
- local to blockchain (in the background, the primary hot wallet will be used to send the funds on the blockchain, but the balance will be deducted from the local wallet)
- blockchain to local (the funds will be received in the hot wallet and will credit the appropriate internal wallet, default to the master account primary wallet)

Reminder of the pooled model:
Local Wallet A ----> Funds pooled into Blockchain Wallet X
Local Wallet B ---/
Local Wallet C--/